packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -1,242 +1,265 @@
+    addDetectedTokens,
+    getTokensState,
+    getPreferencesState,
+    messenger,
+  }: {
+    chainId: Hex;
+    selectedAddress?: string;
+    interval?: number;
+    onPreferencesStateChange: (
+      listener: (preferencesState: PreferencesState) => void,
+    ) => void;
+    getBalancesInSingleCall: AssetsContractController['getBalancesInSingleCall'];
+    addDetectedTokens: TokensController['addDetectedTokens'];
+    getTokensState: () => TokensState;
+    getPreferencesState: () => PreferencesState;
+    messenger: TokenDetectionControllerMessenger;
+  }) {
     const { useTokenDetection: defaultUseTokenDetection } =
       getPreferencesState();
 
-    super(config, state);
-    this.defaultConfig = {
-      interval: DEFAULT_INTERVAL,
-      selectedAddress: '',
-      disabled: true,
-      chainId: defaultChainId,
-      isDetectionEnabledFromPreferences: defaultUseTokenDetection,
-      isDetectionEnabledForNetwork:
-        isTokenDetectionSupportedForNetwork(defaultChainId),
-      ...config,
-    };
-
-    this.initialize();
-    this.setIntervalLength(this.config.interval);
-    this.getTokensState = getTokensState;
-    this.getTokenListState = getTokenListState;
-    this.addDetectedTokens = addDetectedTokens;
-    this.getBalancesInSingleCall = getBalancesInSingleCall;
-    this.getNetworkClientById = getNetworkClientById;
-
-    onTokenListStateChange(({ tokenList }) => {
-      const hasTokens = Object.keys(tokenList).length;
-
-      if (hasTokens) {
-        this.detectTokens();
-      }
+    super({
+      name: controllerName,
+      messenger,
+      state: {},
+      metadata: {},
     });
 
-    onPreferencesStateChange(({ selectedAddress, useTokenDetection }) => {
-      const {
-        selectedAddress: currentSelectedAddress,
-        isDetectionEnabledFromPreferences,
-      } = this.config;
-      const isSelectedAddressChanged =
-        selectedAddress !== currentSelectedAddress;
-      const isDetectionChangedFromPreferences =
-        isDetectionEnabledFromPreferences !== useTokenDetection;
-
-      this.configure({
-        isDetectionEnabledFromPreferences: useTokenDetection,
-        selectedAddress,
-      });
+    this.setIntervalLength(interval);
+    this.#chainId = chainId;
+    this.#selectedAddress = selectedAddress;
 
-      if (
-        useTokenDetection &&
-        (isSelectedAddressChanged || isDetectionChangedFromPreferences)
-      ) {
-        this.detectTokens();
-      }
-    });
+    this.#getTokensState = getTokensState;
+    this.#addDetectedTokens = addDetectedTokens;
+    this.#getBalancesInSingleCall = getBalancesInSingleCall;
 
-    onNetworkStateChange(({ providerConfig: { chainId } }) => {
-      const { chainId: currentChainId } = this.config;
-      const isDetectionEnabledForNetwork =
-        isTokenDetectionSupportedForNetwork(chainId);
-      const isChainIdChanged = currentChainId !== chainId;
+    this.messagingSystem.unregisterActionHandler(
```

Comment:
Hmm, why remove this action handler? Consumers should still be able to access the state of this controller even though it is empty.

Reply:
See https://github.com/MetaMask/core/pull/3609#discussion_r1425686688